src/Microsoft.OData.Client/Metadata/ClientTypeUtil.cs

```diff
@@ -556,27 +569,46 @@ internal static string GetServerDefinedName(MemberInfo memberInfo)
         /// <returns>The server defined type name.</returns>
         internal static string GetServerDefinedTypeName(Type type)
         {
-            OriginalNameAttribute originalNameAttribute = (OriginalNameAttribute)type.GetCustomAttributes(typeof(OriginalNameAttribute), false).SingleOrDefault();
-            if (originalNameAttribute != null)
+            ODataTypeInfo typeInfo = GetODataTypeInfo(type);
+
+            if (typeInfo.ServerDefinedTypeName == null)
             {
-                return originalNameAttribute.OriginalName;
+                OriginalNameAttribute originalNameAttribute = (OriginalNameAttribute)type.GetCustomAttributes(typeof(OriginalNameAttribute), false).SingleOrDefault();
+                if (originalNameAttribute != null)
+                {
+                    typeInfo.ServerDefinedTypeName = originalNameAttribute.OriginalName;
+                }
+                else
+                {
+                    typeInfo.ServerDefinedTypeName = type.Name;
+                }
             }
 
-            return type.Name;
+            return typeInfo.ServerDefinedTypeName;
         }
 
+
         /// <summary>Gets the full server defined type name in <see cref="OriginalNameAttribute"/> of the specified <paramref name="type"/>.</summary>
         /// <param name="type">Member to get server defined name of.</param>
         /// <returns>The server defined type full name.</returns>
         internal static string GetServerDefinedTypeFullName(Type type)
         {
-            OriginalNameAttribute originalNameAttribute = (OriginalNameAttribute)type.GetCustomAttributes(typeof(OriginalNameAttribute), false).SingleOrDefault();
-            if (originalNameAttribute != null)
+            ODataTypeInfo typeInfo = GetODataTypeInfo(type);
+
+            if (typeInfo.ServerDefinedTypeFullName == null)
             {
-                return type.Namespace + "." + originalNameAttribute.OriginalName;
+                OriginalNameAttribute originalNameAttribute = (OriginalNameAttribute)type.GetCustomAttributes(typeof(OriginalNameAttribute), false).SingleOrDefault();
+                if (originalNameAttribute != null)
+                {

```

Comment:
Maybe call GetServiceDefinedTypeName and use it to compute ServerDefinedTypeFullName?